cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaQuorumCheck.java

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright Strimzi authors.
+ * License: Apache License 2.0 (see the file LICENSE or http://apache.org/licenses/LICENSE-2.0.html).
+ */
+package io.strimzi.operator.cluster.operator.resource;
+
+import io.strimzi.operator.common.Reconciliation;
+import io.strimzi.operator.common.ReconciliationLogger;
+import io.strimzi.operator.common.VertxUtil;
+import io.vertx.core.Future;
+import io.vertx.core.Vertx;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.Collectors;
+import org.apache.kafka.clients.admin.Admin;
+import org.apache.kafka.clients.admin.QuorumInfo;
+
+import static java.lang.Math.ceil;
+
+/**
+ * Provides a method that determines whether it's safe to restart a KRaft controller, which may be the active controller
+ */
+class KafkaQuorumCheck {
+
+    private static final ReconciliationLogger LOGGER = ReconciliationLogger.create(KafkaQuorumCheck.class.getName());
+    private final Reconciliation reconciliation;
+    private final Admin admin;
+    private final Vertx vertx;
+    private final long controllerQuorumFetchTimeoutMs;
+    private Future<QuorumInfo> quorumInfoFuture;
+
+    KafkaQuorumCheck(Reconciliation reconciliation, Admin ac, Vertx vertx, long controllerQuorumFetchTimeoutMs) {
+        this.reconciliation = reconciliation;
+        this.admin = ac;
+        this.vertx = vertx;
+        this.controllerQuorumFetchTimeoutMs = controllerQuorumFetchTimeoutMs;
+    }
+
+    /**
+     * Returns future that completes with true if the given controller can be rolled based on the quorum state. Quorum is considered
+     * healthy if the majority of controllers have caught up with the quorum leader within the controller.quorum.fetch.timeout.ms.
+     */
+    Future<Boolean> canRollController(int podId) {
+        LOGGER.debugCr(reconciliation, "Determining whether controller {} can be rolled", podId);
+        if (quorumInfoFuture == null) {
+            this.quorumInfoFuture = describeMetadataQuorum();
+        }
```

Comment:
We used this in 2 checks that are performed subsequently for a pod, one gets leader id and the other one gets the  lastCaughtUpTimestamps.  Thinking about this again, it is better to call it directly, in each checks, rather than reusing it so that we always act on the latest information.